### PR TITLE
adding an atime test case; groupby with dogroups (R expression) #PR4558

### PR DESCRIPTION
**Pull request by DorisAmoakohene**
This test case discusses the issue reported on performing group computations, specifically when running R's C eval on each group (q7 and q8) in the db-benchmark, indicating a slowness in the implementation of the code. https://github.com/Rdatatable/data.table/issues/4200

This is the https://github.com/Rdatatable/data.table/pull/4558 that discusses the Cause of the Regression: https://github.com/Rdatatable/data.table/issues/4200#issue-555186870 which appears that the regression occurred during the evaluation of C code within these particular groups, indicating a performance issue or slowness in the implementation of the code.

[The regression was fixed Regression by the addition of const int nth = getDTthreads](https://github.com/Rdatatable/data.table/pull/4558/files)

### CODE REVIEW

File: .ci/atime/tests.R [hunk 1]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
       data.table:::setDT(L)
     },
     Slow = "c4a2085e35689a108d67dacb2f8261e4964d7e12", # Parent of the first commit in the PR that fixes the issue (https://github.com/Rdatatable/data.table/commit/7cc4da4c1c8e568f655ab5167922dcdb75953801)
-    Fast = "1872f473b20fdcddc5c1b35d79fe9229cd9a1d15") # Last commit in the PR that fixes the issue (https://github.com/Rdatatable/data.table/pull/5427/commits)
+    Fast = "1872f473b20fdcddc5c1b35d79fe9229cd9a1d15"), # Last commit in the PR that fixes the issue (https://github.com/Rdatatable/data.table/pull/5427/commits)
+
+  # Issue reported in: https://github.com/Rdatatable/data.table/issues/4200
+  # To be fixed in: https://github.com/Rdatatable/data.table/pull/4558
+  "groupby with dogroups #4558" = atime::atime_test(
```
Comment (tdhock) [inline .ci/atime/tests.R:126]: for the test name please use "fun improved in #NUMBER" or "fun fixed in #NUMBER" where fun is the name of a data.table function, here I believe fun should be DT[by]

**Comment by DorisAmoakohene** [comment]
@tdhock, 6288 is the right one, when I trying renaming the branch it create a new PR, I think that's the reason for the duplicate, but I am still figuring how I can rename the branch without duplicating the test

File: .ci/atime/tests.R [hunk 2]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+    N = 10^seq(1, 7),
+    setup = {
+      d <- data.table(
+      id3 = sample(c(seq.int(N*0.9), sample(N*0.9, N*0.1, TRUE))),
```
Comment (tdhock) [inline .ci/atime/tests.R:130]: please use consistent indentation, two more spaces after each open paren/brace
two spaces are missing here after the opening data.table( above

File: .ci/atime/tests.R [hunk 3]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+      v1 = sample(5L, N, TRUE),
+      v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
+    },
+    Before="793f8545c363d222de18ac892bc7abb80154e724",#parent of the PR that introduced the regression(https://github.com/Rdatatable/data.table/commit/4aadde8f5a51cd7c8f3889964e7280432ec65bbc) as stated here (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420) https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724
```
Comment (tdhock) [inline .ci/atime/tests.R:138]: please use consistent spacing and capitalization (see examples above) `", # Parent`

Comment (tdhock) [reply .ci/atime/tests.R:138]: please put commit link after the word commit in your comment
Parent of the first commit (LINK) of the PR (LINK) that introduced the regression

File: .ci/atime/tests.R [hunk 4]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+    Regression="c152ced0e5799acee1589910c69c1a2c6586b95d", #praent of the first commit in the PR (https://github.com/Rdatatable/data.table/commit/15f0598b9828d3af2eb8ddc9b38e0356f42afe4f)
```
Comment (tdhock) [inline .ci/atime/tests.R:139]: typo -> Parent of the first commit (LINK) of the PR (LINK) that fixed the regression?

File: .ci/atime/tests.R [hunk 5]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+    Fixed="f750448a2efcd258b3aba57136ee6a95ce56b302")#second commit in the PR that fixes the regression(https://github.com/Rdatatable/data.table/pull/4558/commits)
```
Comment (tdhock) [inline .ci/atime/tests.R:140]: Second commit (LINK) of the PR (LINK) that fixed the regression

**Comment by DorisAmoakohene** [comment]
@tdhock, Kindly see the updates, so I can continue with the other PR

**Comment by tdhock** [comment]
Both commits in this PR have message "Update tests.R"
Please use 
```
git commit -m "More specific commit message which is a short description of the changes that you made"
```

File: .ci/atime/tests.R (hunk 3 above)
Comment (tdhock) [reply .ci/atime/tests.R]: please use comment template below
Parent of the first commit (LINK) of the PR (LINK) that introduced the regression

File: .ci/atime/tests.R [hunk 6]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+      d <- data.table(
+        id3 = sample( c(seq.int(N*0.9), sample( N*0.9, N*0.1, TRUE))),
+        v1 = sample(5L, N, TRUE),
+        v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
+    },
+    Before="793f8545c363d222de18ac892bc7abb80154e724", # Parent of the PR that introduced the regression(https://github.com/Rdatatable/data.table/commit/4aadde8f5a51cd7c8f3889964e7280432ec65bbc) as stated here (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420) https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724
```
Comment (tdhock) [inline .ci/atime/tests.R:138]: Which commit in the PR? Please answer by using the template below 
Parent of the first commit (LINK) of the PR (LINK) that introduced the regression

Comment (tdhock) [reply .ci/atime/tests.R:138]: word(LINK) -> word (LINK)
please add spaces before and after parens

File: .ci/atime/tests.R [hunk 7]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+    Regression="c152ced0e5799acee1589910c69c1a2c6586b95d", # Parent of the first commit(https://github.com/Rdatatable/data.table/pull/4558/commits/15f0598b9828d3af2eb8ddc9b38e0356f42afe4f) in the PR (https://github.com/Rdatatable/data.table/pull/4558/commits)
```
Comment (tdhock) [inline .ci/atime/tests.R:139]: please read https://github.com/Rdatatable/data.table/wiki/Performance-testing#documentation-of-historical-commits which explains that commit (LINK) should not be data.table/pull/NUMBER/commits/ID but instead data.table/commit/ID
"do not include a link which includes data.table/pull/number/commits/sha (does not show parent) instead use data.table/commit/sha because that page shows the parent"

File: .ci/atime/tests.R (hunk 1 above)
Comment (DorisAmoakohene) [reply .ci/atime/tests.R]: This has been addressed

File: .ci/atime/tests.R (hunk 2 above)
Comment (DorisAmoakohene) [reply .ci/atime/tests.R]: Addressed

File: .ci/atime/tests.R (hunk 6 above)
Comment (DorisAmoakohene) [reply .ci/atime/tests.R:138]: Okay

File: .ci/atime/tests.R [hunk 8]
```diff
@@ -108,7 +108,7 @@ test.list <- atime::atime_test_list(
 
   # Issue reported in: https://github.com/Rdatatable/data.table/issues/5426
   # To be fixed in: https://github.com/Rdatatable/data.table/pull/5427
-  "setDT improved in #5427" = atime::atime_test(
+  "setDT improved fixed in #5427" = atime::atime_test(
```
Comment (tdhock) [inline .ci/atime/tests.R:111]: please undo this change as it is unrelated to the test you are adding in this PR

File: .ci/atime/tests.R [hunk 9]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+      d <- data.table(
+        id3 = sample( c(seq.int(N*0.9), sample( N*0.9, N*0.1, TRUE))),
+        v1 = sample(5L, N, TRUE),
+        v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
+    },
+    Before="793f8545c363d222de18ac892bc7abb80154e724", # This is the link to the Before commit (https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724) as stated here (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420)
```
Comment (tdhock) [inline .ci/atime/tests.R:138]: please avoid "This is the link to" which does not add any information.
instead please use comment template https://github.com/Rdatatable/data.table/wiki/Performance-testing#documentation-of-historical-commits

File: .ci/atime/tests.R [hunk 10]
```diff
@@ -135,9 +135,9 @@ test.list <- atime::atime_test_list(
     expr = {
       expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
     },
-    Before="793f8545c363d222de18ac892bc7abb80154e724", # This is the link to the Before commit (https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724) as stated here (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420) 
+    Before="793f8545c363d222de18ac892bc7abb80154e724", # Commit (https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724) which was Before the regression was introduced as stated in the issue (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420)
```
Comment (tdhock) [inline .ci/atime/tests.R:138]: this is still not an acceptable comment, because "do not include the link to the parent commit https://github.com/Rdatatable/data.table/commit/cacdc92df71b777369a217b6c902c687cf35a70d as that does not help understanding how the commit is related to the PR/issue, and it is redundant with the code which already has that ID."
Please read all of https://github.com/Rdatatable/data.table/wiki/Performance-testing#documentation-of-historical-commits and write a comment which follows those rules, with minimal deviation from the templates.

Comment (tdhock) [reply .ci/atime/tests.R:138]: I revised that guideline to clarify:

do not include the link to the same commit ID as in the code, as that does not help understanding how the commit is related to the PR/issue, and it is redundant with the code which already has that ID. For example the code above has `Slow = "cacdc92df71b777369a217b6c902c687cf35a70d"` so it is not necessary to inclde the link https://github.com/Rdatatable/data.table/commit/cacdc92df71b777369a217b6c902c687cf35a70d

**Comment by tdhock** [comment]
hi @Anirban166 can you please help by reviewing and giving suggestions about how to improve this PR?

File: .ci/atime/tests.R (hunk 10 above)
Comment (DorisAmoakohene) [reply .ci/atime/tests.R]: I have updated the Before commit id and given an appropriate comment , please check

File: .ci/atime/tests.R (hunk 8 above)
Comment (Anirban166) [inline .ci/atime/tests.R:111]: Please revert the change here (the altered wording sounds wrong and this is a separate test)

File: .ci/atime/tests.R [hunk 11]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
-    Fast = "1872f473b20fdcddc5c1b35d79fe9229cd9a1d15") # Last commit in the PR that fixes the issue (https://github.com/Rdatatable/data.table/pull/5427/commits)
+    Fast = "1872f473b20fdcddc5c1b35d79fe9229cd9a1d15"), # Last commit in the PR that fixes the issue (https://github.com/Rdatatable/data.table/pull/5427/commits)
+
+  # Issue reported in: https://github.com/Rdatatable/data.table/issues/4200
+  # To be fixed in: https://github.com/Rdatatable/data.table/pull/4558
+  "DT[by] fixed in #4558" = atime::atime_test(
+    N = 10^seq(1, 20),
+    setup = {
+      d <- data.table(
+        id3 = sample( c(seq.int(N*0.9), sample( N*0.9, N*0.1, TRUE))),
```
Comment (Anirban166) [inline .ci/atime/tests.R:130]: Remove the spaces after `sample(` here

File: .ci/atime/tests.R [hunk 12]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+        v1 = sample(5L, N, TRUE),
+        v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
+    },
+    Before="905e4d302be0dc326684d7228ece88064321cc00", # Parent of the regression commit (https://github.com/Rdatatable/data.table/commit/c152ced0e5799acee1589910c69c1a2c6586b95d)
```
Comment (Anirban166) [inline .ci/atime/tests.R:138]: Spaces before and after the `=` (lines 136, 138, 139, 140) would be nice (and perhaps for operators like `-` and `*` above too you can have spacing before and after)

File: .ci/atime/tests.R [hunk 13]
```diff
@@ -119,6 +119,25 @@ test.list <- atime::atime_test_list(
...
+    Regression="c152ced0e5799acee1589910c69c1a2c6586b95d", # Parent of the first commit (https://github.com/Rdatatable/data.table/commit/15f0598b9828d3af2eb8ddc9b38e0356f42afe4f) in the PR (https://github.com/Rdatatable/data.table/pull/4558/commits) that fixes the regression
+    Fixed="f750448a2efcd258b3aba57136ee6a95ce56b302") # Second commit (https://github.com/Rdatatable/data.table/commit/f750448a2efcd258b3aba57136ee6a95ce56b302) in the PR  that fixes the regression(https://github.com/Rdatatable/data.table/pull/4558/commits)
```
Comment (Anirban166) [inline .ci/atime/tests.R:140]: Consistent spacing again please - remove the extra one between 'PR' and 'that', and add one between 'regression' and '('

File: .ci/atime/tests.R (hunk 12 above)
Comment (tdhock) [reply .ci/atime/tests.R:138]: I agree. You should look at the nearby code and try to use the same style.

Comment (tdhock) [reply .ci/atime/tests.R:138]: also parenthesis around max-min are not necessary, please remove

File: .ci/atime/tests.R [hunk 14]
```diff
@@ -135,7 +135,7 @@ test.list <- atime::atime_test_list(
     expr = {
       expr=data.table:::`[.data.table`(d, , (max(v1)-min(v2)), by = id3)
     },
-    Before="793f8545c363d222de18ac892bc7abb80154e724", # Commit (https://github.com/Rdatatable/data.table/commit/793f8545c363d222de18ac892bc7abb80154e724) which was Before the regression was introduced as stated in the issue (https://github.com/Rdatatable/data.table/issues/4200#issuecomment-646111420) 
+    Before="905e4d302be0dc326684d7228ece88064321cc00", # Parent of the regression commit (https://github.com/Rdatatable/data.table/commit/c152ced0e5799acee1589910c69c1a2c6586b95d)
```
Comment (tdhock) [inline .ci/atime/tests.R:138]: why did you change the Before commit ID?

The Before commit comment you wrote is "Parent of the regression commit (https://github.com/Rdatatable/data.table/commit/c152ced0e5799acee1589910c69c1a2c6586b95d)" which is not consistent with the suggested template, "Parent of first commit (COMMIT_LINK) in the PR (PR_LINK) which causes the issue" 

please re-read https://github.com/Rdatatable/data.table/wiki/Performance-testing#documentation-of-historical-commits and if there is anything that you do not understand, please ask for clarification.

**Comment by DorisAmoakohene** [comment]
Hi @tdhock  and @Anirban166 could you review this for merge.

**Comment by Anirban166** [comment]
There are still pending reviews/comments that you need to push changes for

**Comment by DorisAmoakohene** [comment]
> There are still pending reviews/comments that you need to push changes for

I was supposed to update the comment for the Commit id, and I have

**Comment by tdhock** [comment]
please click Files Changed tab, and read comments, then push commit(s) that address those comments.

File: .ci/atime/tests.R (hunk 8 above)
Comment (tdhock) [reply .ci/atime/tests.R:111]: this is not resolved, please change "improved fixed" back to "improved"

File: .ci/atime/tests.R [hunk 15]
```diff
@@ -119,6 +119,24 @@ test.list <- atime::atime_test_list(
...
+        v1 = sample(5L, N, TRUE),
+        v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , max(v1) - min(v2), by = id3)
+    },
+    Before = "7a9eaf62ede487625200981018d8692be8c6f134", # Parent of the first commit (https://github.com/Rdatatable/data.table/commit/515de90a6068911a148e54343a3503043b8bb87c) in the PR (https://github.com/Rdatatable/data.table/pull/4164/commits) that introduced the regression
+    Regression = "c152ced0e5799acee1589910c69c1a2c6586b95d", # Parent of the first (https://github.com/Rdatatable/data.table/commit/15f0598b9828d3af2eb8ddc9b38e0356f42afe4f) in the PR (https://github.com/Rdatatable/data.table/pull/4558/commits) that fixes the regression
+    Fixed = "f750448a2efcd258b3aba57136ee6a95ce56b302") # Second commit of the PR that fixes the regression(https://github.com/Rdatatable/data.table/pull/4558/commits)
```
Comment (tdhock) [inline .ci/atime/tests.R:140]: please use space before open paren
bad: regression(http
good: regression (http

File: .ci/atime/tests.R [hunk 16]
```diff
@@ -119,6 +119,24 @@ test.list <- atime::atime_test_list(
...
+        id3 = sample(c(seq.int(N*0.9), sample( N*0.9, N*0.1, TRUE))),
+        v1 = sample(5L, N, TRUE),
+        v2 = sample(5L, N, TRUE)
+      )
+    },
+    expr = {
+      expr=data.table:::`[.data.table`(d, , max(v1) - min(v2), by = id3)
+    },
+    Before = "7a9eaf62ede487625200981018d8692be8c6f134", # Parent of the first commit (https://github.com/Rdatatable/data.table/commit/515de90a6068911a148e54343a3503043b8bb87c) in the PR (https://github.com/Rdatatable/data.table/pull/4164/commits) that introduced the regression
+    Regression = "c152ced0e5799acee1589910c69c1a2c6586b95d", # Parent of the first (https://github.com/Rdatatable/data.table/commit/15f0598b9828d3af2eb8ddc9b38e0356f42afe4f) in the PR (https://github.com/Rdatatable/data.table/pull/4558/commits) that fixes the regression
```
Comment (tdhock) [inline .ci/atime/tests.R:139]: please add "commit" : parent of the first commit

File: .ci/atime/tests.R (hunk 15 above)
Comment (tdhock) [reply .ci/atime/tests.R:140]: and actually please move the URL immediately after PR
good: PR (URL)
bad: regression (URL)
